### PR TITLE
Potion NamespacedKey

### PR DESCRIPTION
**Pull request by DaRacci**
Unlike the previous api change i had made a PR for this time it _actually works_ and is now utilizing the namespacedkey system.

### CODE REVIEW

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
++    private NamespacedKey key;
+     @Nullable
+     private MobEffectInstance hiddenEffect;
+ 
+@@ -36,16 +38,26 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
+     }
+ 
+     public MobEffectInstance(MobEffect type, int duration, int amplifier, boolean ambient, boolean showParticles, boolean showIcon) {
+-        this(type, duration, amplifier, ambient, showParticles, showIcon, (MobEffectInstance)null);
++        this(type, duration, amplifier, ambient, showParticles, showIcon, null);
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: unnecessary diff

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
++        }
++        // Purpur end
+         if (this.hiddenEffect != null) {
+             CompoundTag compoundTag = new CompoundTag();
+             this.hiddenEffect.save(compoundTag);
+@@ -247,13 +287,14 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
+         if (nbt.contains("ShowIcon", 1)) {
+             bl3 = nbt.getBoolean("ShowIcon");
+         }
++        NamespacedKey key = NamespacedKey.fromString(nbt.getString("Key"));
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: add Purpur comment

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+ 
+         MobEffectInstance mobEffectInstance = null;
+         if (nbt.contains("HiddenEffect", 10)) {
+             mobEffectInstance = loadSpecifiedEffect(type, nbt.getCompound("HiddenEffect"));
+         }
+ 
+-        return new MobEffectInstance(type, j, i < 0 ? 0 : i, bl, bl2, bl3, mobEffectInstance);
++        return new MobEffectInstance(type, j, i < 0 ? 0 : i, bl, bl2, bl3, key, mobEffectInstance);
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+     }
+ 
+     public static PotionEffect toBukkit(MobEffectInstance effect) {
+@@ -110,7 +111,8 @@ public class CraftPotionUtil {
+         int duration = effect.getDuration();
+         boolean ambient = effect.isAmbient();
+         boolean particles = effect.isVisible();
+-        return new PotionEffect(type, duration, amp, ambient, particles);
++        NamespacedKey key = effect.getKey(); // Purpur - add key
++        return new PotionEffect(type, duration, amp, ambient, particles, key);
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: No need to create a variable, just return directly :
` return new PotionEffect(type, duration, amp, ambient, particles, effect.getKey()); // Purpur`

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                 this.customEffects.set(index, effect);
+diff --git a/src/main/java/org/bukkit/craftbukkit/potion/CraftPotionUtil.java b/src/main/java/org/bukkit/craftbukkit/potion/CraftPotionUtil.java
+index acb69821a99aa69bce6d127e10976089c85be223..01bf7719f0b2d85d385915cbd5f18df341eec727 100644
+--- a/src/main/java/org/bukkit/craftbukkit/potion/CraftPotionUtil.java
++++ b/src/main/java/org/bukkit/craftbukkit/potion/CraftPotionUtil.java
+@@ -5,6 +5,7 @@ import com.google.common.collect.BiMap;
+ import com.google.common.collect.ImmutableBiMap;
+ import net.minecraft.world.effect.MobEffect;
+ import net.minecraft.world.effect.MobEffectInstance;
++import org.bukkit.NamespacedKey;
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: unnecessary diff

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+ import org.apache.logging.log4j.Logger;
++import org.bukkit.NamespacedKey;
+ 
+ public class MobEffectInstance implements Comparable<MobEffectInstance> {
+     private static final Logger LOGGER = LogManager.getLogger();
+@@ -16,6 +17,7 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
+     private boolean noCounter;
+     private boolean visible;
+     private boolean showIcon;
++    private NamespacedKey key;
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: Add Purpur comment

File: patches/api/0050-Potion-NamespacedKey.patch [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
++    // Purpur end
++
+     @NotNull
+     private static PotionEffectType getEffectType(@NotNull Map<?, ?> map) {
+         int type = getInt(map, TYPE);
+@@ -154,6 +196,14 @@ public class PotionEffect implements ConfigurationSerializable {
+         return def;
+     }
+ 
++    private static NamespacedKey getKey(@NotNull Map<?, ?> map) {
```
Comment (Euphillya) [inline patches/api/0050-Potion-NamespacedKey.patch:113]: Add Purpur comment

File: patches/api/0050-Potion-NamespacedKey.patch [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
++     * @param type effect type
++     * @param duration measured in ticks, see {@link
++     *     PotionEffect#getDuration()}
++     * @param amplifier the amplifier, see {@link PotionEffect#getAmplifier()}
++     * @param ambient the ambient status, see {@link PotionEffect#isAmbient()}
++     * @param particles the particle status, see {@link PotionEffect#hasParticles()}
++     * @param icon the icon status, see {@link PotionEffect#hasIcon()}
++     */
++    public PotionEffect(@NotNull PotionEffectType type, int duration, int amplifier, boolean ambient, boolean particles, boolean icon) {
++        this(type, duration, amplifier, ambient, particles, icon, null);
```
Comment (Euphillya) [inline patches/api/0050-Potion-NamespacedKey.patch:79]: Add Purpur End comment

File: patches/api/0050-Potion-NamespacedKey.patch [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+ import org.bukkit.Color;
++import org.bukkit.NamespacedKey;
+ import org.bukkit.configuration.serialization.ConfigurationSerializable;
+ import org.bukkit.configuration.serialization.SerializableAs;
+ import org.bukkit.entity.LivingEntity;
+@@ -26,12 +27,14 @@ public class PotionEffect implements ConfigurationSerializable {
+     private static final String AMBIENT = "ambient";
+     private static final String PARTICLES = "has-particles";
+     private static final String ICON = "has-icon";
++    private static final String KEY = "namespacedkey";
```
Comment (Euphillya) [inline patches/api/0050-Potion-NamespacedKey.patch]: Add Purpur Comment

**Review by Euphillya — CHANGES_REQUESTED** [review]

File: patches/api/0050-Potion-NamespacedKey.patch [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+     /**
+      * Creates a potion effect.
+@@ -42,8 +45,9 @@ public class PotionEffect implements ConfigurationSerializable {
+      * @param ambient the ambient status, see {@link PotionEffect#isAmbient()}
+      * @param particles the particle status, see {@link PotionEffect#hasParticles()}
+      * @param icon the icon status, see {@link PotionEffect#hasIcon()}
++     * @param key the namespacedkey, see {@link PotionEffect#getKey()}
+      */
+-    public PotionEffect(@NotNull PotionEffectType type, int duration, int amplifier, boolean ambient, boolean particles, boolean icon) {
++    public PotionEffect(@NotNull PotionEffectType type, int duration, int amplifier, boolean ambient, boolean particles, boolean icon, NamespacedKey key) {
```
Comment (Euphillya) [inline patches/api/0050-Potion-NamespacedKey.patch]: Add Purpur Comment

**Review by Euphillya — CHANGES_REQUESTED** [review]

File: patches/server/0262-Potion-NamespacedKey.patch (hunk 1 above)
Comment (DaRacci) [reply patches/server/0262-Potion-NamespacedKey.patch]: woops didn't mean to include that

Comment (DaRacci) [reply patches/server/0262-Potion-NamespacedKey.patch]: wait no actually thats needed because i changed the order somewhat hrrr

Comment (DaRacci) [reply patches/server/0262-Potion-NamespacedKey.patch]: would it be better to implement a new constructor to match the old order or leave it?

Comment (Euphillya) [reply patches/server/0262-Potion-NamespacedKey.patch]: I think new constructor is better to maintien API

File: patches/server/0262-Potion-NamespacedKey.patch [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
++        return new MobEffectInstance(type, effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles(), effect.getKey()); // Purpur - add key
+     }
+ 
+     public static PotionEffect toBukkit(MobEffectInstance effect) {
+@@ -110,7 +110,7 @@ public class CraftPotionUtil {
+         int duration = effect.getDuration();
+         boolean ambient = effect.isAmbient();
+         boolean particles = effect.isVisible();
+-        return new PotionEffect(type, duration, amp, ambient, particles);
++        return new PotionEffect(type, duration, amp, ambient, particles, effect.getKey());
```
Comment (Euphillya) [inline patches/server/0262-Potion-NamespacedKey.patch]: Add Purpur comment

**Review by Euphillya — CHANGES_REQUESTED** [review]

**Review by encode42 — COMMENTED** [review]
LGTM, needs a rebase though

**Comment by DaRacci** [comment]
Sorry for closing and opening again, check https://github.com/pl3xgaming/Purpur/pull/754 for the updated PR

**Pull request closed without merging**